### ISSUE TITLE
Memory leak query endpoint (httpClusterListenAddress)

### ISSUE THREAD

**Issue by sboschman** [m3db/m3#869]
M3DB 0.4.1

Using a query like the single node deployment example (http://m3db.github.io/m3/how_to/single_node/) does for reading metrics returns no response and increases memory usage by roughly 500mb for each executed query. Repeatedly issuing queries will oom your server.

---

A freshly started node (entire cluster was restarted) after bootstrap is complete:
```
File: m3dbnode
Type: inuse_space
Time: Aug 29, 2018 at 4:41pm (CEST)
Showing nodes accounting for 1920.23MB, 75.48% of 2544MB total
Dropped 271 nodes (cum <= 12.72MB)
Showing top 5 nodes out of 128
      flat  flat%   sum%        cum   cum%
 1042.62MB 40.98% 40.98%  1042.62MB 40.98%  github.com/m3db/m3/src/dbnode/client.newReceivedBlocksMap.func2
     352MB 13.84% 54.82%      352MB 13.84%  github.com/m3db/m3/src/dbnode/persist/fs/commitlog.NewCommitLog
  254.62MB 10.01% 64.83%   254.62MB 10.01%  testing.init
  193.98MB  7.62% 72.45%   193.98MB  7.62%  github.com/m3db/m3/src/m3ninx/index/segment/mem.(*bytesSliceIter).Close
   77.01MB  3.03% 75.48%    92.53MB  3.64%  github.com/m3db/m3/src/dbnode/serialize.newTagEncoder
```

---

Same node a few hours later, it only had to ingest new metrics:
```
File: m3dbnode
Type: inuse_space
Time: Aug 30, 2018 at 10:55am (CEST)
Showing nodes accounting for 1730.60MB, 70.32% of 2461.01MB total
Dropped 400 nodes (cum <= 12.31MB)
Showing top 5 nodes out of 144
      flat  flat%   sum%        cum   cum%
 1042.62MB 42.37% 42.37%  1042.62MB 42.37%  github.com/m3db/m3/src/dbnode/client.newReceivedBlocksMap.func2
     352MB 14.30% 56.67%      352MB 14.30%  github.com/m3db/m3/src/dbnode/persist/fs/commitlog.NewCommitLog
  193.46MB  7.86% 64.53%   193.46MB  7.86%  github.com/m3db/m3/src/m3ninx/index/segment/mem.(*bytesSliceIter).Close
   77.01MB  3.13% 67.66%    92.53MB  3.76%  github.com/m3db/m3/src/dbnode/serialize.newTagEncoder
   65.50MB  2.66% 70.32%    65.50MB  2.66%  github.com/m3db/m3/vendor/github.com/uber-go/tally.(*bucketPair).LowerBoundValue
```

Node after issuing a query to port 9003 (httpClusterListenAddress):
```
File: m3dbnode
Type: inuse_space
Time: Aug 30, 2018 at 11:20am (CEST)
Showing nodes accounting for 2954.88MB, 77.22% of 3826.76MB total
Dropped 439 nodes (cum <= 19.13MB)
Showing top 5 nodes out of 151
      flat  flat%   sum%        cum   cum%
 1278.79MB 33.42% 33.42%  1278.79MB 33.42%  github.com/m3db/m3/vendor/github.com/m3db/m3x/pool.(*nativePool).Put
 1042.62MB 27.25% 60.66%  1042.62MB 27.25%  github.com/m3db/m3/src/dbnode/client.newReceivedBlocksMap.func2
     352MB  9.20% 69.86%      352MB  9.20%  github.com/m3db/m3/src/dbnode/persist/fs/commitlog.NewCommitLog
  193.46MB  5.06% 74.92%   193.46MB  5.06%  github.com/m3db/m3/src/m3ninx/index/segment/mem.(*bytesSliceIter).Close
   88.01MB  2.30% 77.22%    88.01MB  2.30%  github.com/m3db/m3/vendor/github.com/uber-go/tally.(*bucketPair).LowerBoundValue
```

Issuing the same query again:
```
File: m3dbnode
Type: inuse_space
Time: Aug 30, 2018 at 11:24am (CEST)
Showing nodes accounting for 3522.59MB, 78.74% of 4473.78MB total
Dropped 430 nodes (cum <= 22.37MB)
Showing top 5 nodes out of 153
      flat  flat%   sum%        cum   cum%
 1815.99MB 40.59% 40.59%  1815.99MB 40.59%  github.com/m3db/m3/vendor/github.com/m3db/m3x/pool.(*nativePool).Put
 1042.62MB 23.31% 63.90%  1042.62MB 23.31%  github.com/m3db/m3/src/dbnode/client.newReceivedBlocksMap.func2
     352MB  7.87% 71.77%      352MB  7.87%  github.com/m3db/m3/src/dbnode/persist/fs/commitlog.NewCommitLog
  193.46MB  4.32% 76.09%   193.46MB  4.32%  github.com/m3db/m3/src/m3ninx/index/segment/mem.(*bytesSliceIter).Close
  118.51MB  2.65% 78.74%   118.51MB  2.65%  github.com/m3db/m3/vendor/github.com/uber-go/tally.(*bucketPair).LowerBoundValue
```

Let's issue it again!
```
File: m3dbnode
Type: inuse_space
Time: Aug 30, 2018 at 11:27am (CEST)
Showing nodes accounting for 4108.07MB, 79.87% of 5143.44MB total
Dropped 434 nodes (cum <= 25.72MB)
Showing top 5 nodes out of 152
      flat  flat%   sum%        cum   cum%
 2375.97MB 46.19% 46.19%  2375.97MB 46.19%  github.com/m3db/m3/vendor/github.com/m3db/m3x/pool.(*nativePool).Put
 1042.62MB 20.27% 66.47%  1042.62MB 20.27%  github.com/m3db/m3/src/dbnode/client.newReceivedBlocksMap.func2
     352MB  6.84% 73.31%      352MB  6.84%  github.com/m3db/m3/src/dbnode/persist/fs/commitlog.NewCommitLog
  193.46MB  3.76% 77.07%   193.46MB  3.76%  github.com/m3db/m3/src/m3ninx/index/segment/mem.(*bytesSliceIter).Close
  144.01MB  2.80% 79.87%   144.01MB  2.80%  github.com/m3db/m3/vendor/github.com/uber-go/tally.(*bucketPair).LowerBoundValue
```

Details for Put:
```
ROUTINE ======================== github.com/m3db/m3/vendor/github.com/m3db/m3x/pool.(*nativePool).Put in src/github.com/m3db/m3/vendor/github.com/m3db/m3x/pool/native.go
    2.32GB     2.32GB (flat, cum) 46.19% of Total
         .          .    134:           return fn()
         .          .    135:   }
         .          .    136:}
         .          .    137:
         .          .    138:// Put returns an object to the pool.
    2.32GB     2.32GB    139:func (p *nativePool) Put(object interface{}) {
    4.50MB     4.50MB    140:   ptr := unsafe.Pointer(reflect.ValueOf(object).Pointer())
         .          .    141:
         .          .    142:   if !p.owns(ptr) {
         .          .    143:           return
         .          .    144:   }
         .          .    145:
```

Snippet from the config:
```
  pooling:
    blockAllocSize: 16
    type: simple
```

Snippet from the log:
```
09:52:21.758317[W] max index query IDs concurrency was not set, falling back to default value
09:52:21.758628[W] host doesn't support HugeTLB, proceeding without it
09:52:21.779076[I] bytes pool registering bucket capacity=16, size=52428, refillLowWatermark=0.700000, refillHighWatermark=1.000000
09:52:21.779098[I] bytes pool registering bucket capacity=32, size=26214, refillLowWatermark=0.700000, refillHighWatermark=1.000000
09:52:21.779103[I] bytes pool registering bucket capacity=64, size=13107, refillLowWatermark=0.700000, refillHighWatermark=1.000000
09:52:21.779111[I] bytes pool registering bucket capacity=128, size=6553, refillLowWatermark=0.700000, refillHighWatermark=1.000000
09:52:21.779115[I] bytes pool registering bucket capacity=256, size=6553, refillLowWatermark=0.700000, refillHighWatermark=1.000000
09:52:21.779119[I] bytes pool registering bucket capacity=1440, size=1638, refillLowWatermark=0.700000, refillHighWatermark=1.000000
09:52:21.779125[I] bytes pool registering bucket capacity=4096, size=819, refillLowWatermark=0.700000, refillHighWatermark=1.000000
09:52:21.779175[I] bytes pool simple init
```

**Comment by robskillington** [comment]
Hey @sboschman thanks for the detailed report.

It's very strange that the native pool is being used, even the logs register that you're using simple pooling as evidenced by the log message `bytes pool simple init`.

I'm going to trace down why that endpoint is using the native pool.  

For now I'm disabling the use of the native pool in config too:
https://github.com/m3db/m3/pull/870

**Comment by sboschman** [comment]
Cluster init config I used:
```
curl -X POST localhost:7201/api/v1/placement/init -d '{
    "num_shards": 256,
    "replication_factor": 2,
    "instances": [
        {
            "id": "m3db-node01",
            "isolation_group": "group1",
            "zone": "embedded",
            "weight": 100,
            "endpoint": "192.168.1.1:9000",
            "hostname": "m3db-node01",
            "port": 9000
        },
        {
            "id": "m3db-node02",
            "isolation_group": "group1",
            "zone": "embedded",
            "weight": 100,
            "endpoint": "192.168.1.2:9000",
            "hostname": "m3db-node02",
            "port": 9000
        },
        {
            "id": "m3db-node03",
            "isolation_group": "group2",
            "zone": "embedded",
            "weight": 100,
            "endpoint": "192.168.1.3:9000",
            "hostname": "m3db-node03",
            "port": 9000
        },
        {
            "id": "m3db-node04",
            "isolation_group": "group2",
            "zone": "embedded",
            "weight": 100,
            "endpoint": "192.168.1.4:9000",
            "hostname": "m3db-node04",
            "port": 9000
        }
    ]
}'
```

Namepace init script:
```
curl -X POST localhost:7201/api/v1/namespace -d '{
  "name": "metrics",
  "options": {
    "bootstrapEnabled": true,
    "flushEnabled": true,
    "writesToCommitLog": true,
    "cleanupEnabled": true,
    "snapshotEnabled": false,
    "repairEnabled": false,
    "retentionOptions": {
      "retentionPeriodDuration": "720h",
      "blockSizeDuration": "12h",
      "bufferFutureDuration": "1h",
      "bufferPastDuration": "1h",
      "blockDataExpiry": true,
      "blockDataExpiryAfterNotAccessPeriodDuration": "5m"
    },
    "indexOptions": {
      "enabled": true,
      "blockSizeDuration": "12h"
    }
  }
}'
```

**Comment by robskillington** [comment]
I've completely removed the native pool from both our libraries and the dbnode source code with #870 and cutting a release once the PR is merged.

**Comment by robskillington** [comment]
Hey @sboschman we've cut a release https://github.com/m3db/m3/releases/tag/v0.4.2 which removes all traces of native pooling, would you mind testing it for us whenever is convenient?  

Thanks a lot again for the detailed report.

**Comment by robskillington** [comment]
I believe this was fixed in #870 .
